internal/services/rate_test.go

```diff
@@ -0,0 +1,42 @@
+package services
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+const (
+	TestingRateServiceFetchInterval   = time.Second * 1
+	TestingRateServiceWaitingDuration = time.Second * 2
+)
+
+type MockExchangeRateClient struct {
+	mock.Mock
+}
+
+func (ec *MockExchangeRateClient) GetExchangeRate() (float32, error) {
+	args := ec.Called()
+	return args.Get(0).(float32), args.Error(1)
+}
+
+func TestRateService(t *testing.T) {
```

Comment:
Here you're testing several things at the same time:
- if fetched rate is correct
- if rate is cached
- if rate is re-fetched after update interval

These are different act parts from AAA approach.
I'd suggest implementing it as different tests.
Then tests will become more clear and easier to maintain

Reply:
Resolved with [95ae405](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/4/commits/95ae405cbefc92d7ec1edb2d1f12c7e469b05e9b)